model.py

```diff
@@ -5424,50 +5579,96 @@ def calculate_work(self, even_if_no_author=False, known_edition=None):
-
-        else:
-            _db = Session.object_session(self)
-            work = None
-            if self.open_access and presentation_edition.permanent_work_id:
-                # Is there already an open-access Work which includes editions
-                # with this edition's permanent work ID?
-                q = _db.query(Edition).filter(
-                    Edition.permanent_work_id
-                    ==presentation_edition.permanent_work_id).filter(
-                        Edition.work != None).filter(
-                            Edition.id != presentation_edition.id)
-                for edition in q:
-                    if edition.work.has_open_access_license:
-                        work = edition.work
-                        break
+            self.work = work
 
         if work:
-            created = False
+            # There is already a Work associated with this LicensePool,
+            # but we need to run a sanity check because occasionally
+            # LicensePools get mis-grouped due to bugs.
+            #
+            # A commercially-licensed book should have a Work to
+            # itself. All other LicensePools need to be kicked out and
+            # associated with some other work.
+            #
+            # This won't cause an infinite recursion because we're
+            # setting pool.work to None before calling
+            # pool.calculate_work(), and the recursive call only
+            # happens if self.work is set.
+            my_pwid = self.presentation_edition
+            for pool in list(work.license_pools):
+                if pool is self:
+                    continue
+                if not (self.open_access and pool.open_access):
+                    pool.work = None
+                    pool.calculate_work()
+                    licensepools_changed = True
+
         else:
             # There is no better choice than creating a brand new Work.
-            created = True
-            logging.info("NEW WORK for %s" % presentation_edition.title)
+            is_new = True
+            logging.info(
+                "Creating a new work for %r" % presentation_edition.title
+            )
             work = Work()
             _db = Session.object_session(self)
             _db.add(work)
             _db.flush()
+            licensepools_changed = True
 
         # Associate this LicensePool and its Edition with the work we
         # chose or created.
-        work.license_pools.append(self)
+        if not self in work.license_pools:
+            work.license_pools.append(self)
+            licensepools_changed = True
 
         # Recalculate the display information for the Work, since the
-        # associated Editions have changed.
-        work.calculate_presentation()
+        # associated LicensePools have changed, which may have caused
+        # the Work's presentation Edition to change.
+        if licensepools_changed:
+            work.calculate_presentation()
```

Comment:
I've been assuming calculate_work always calls calculate_presentation, so there may be some places we need to add a call to calculate_presentation now.